sklearn/neighbors/_kde.py

```diff
@@ -304,22 +303,17 @@ def sample(self, n_samples=1, random_state=None):
             cumsum_weight = np.cumsum(np.asarray(self.tree_.sample_weight))
             sum_weight = cumsum_weight[-1]
             i = np.searchsorted(cumsum_weight, u * sum_weight)
-        if self.kernel == "gaussian":
-            return np.atleast_2d(rng.normal(data[i], self.bandwidth))
 
+        size = data[i].shape
+        if self.kernel == "gaussian":
+            additional = rng.normal(0, self.bandwidth, size=size)
         elif self.kernel == "tophat":
-            # we first draw points from a d-dimensional normal distribution,
-            # then use an incomplete gamma function to map them to a uniform
-            # d-dimensional tophat distribution.
-            dim = data.shape[1]
-            X = rng.normal(size=(n_samples, dim))
-            s_sq = row_norms(X, squared=True)
-            correction = (
-                gammainc(0.5 * dim, 0.5 * s_sq) ** (1.0 / dim)
-                * self.bandwidth
-                / np.sqrt(s_sq)
-            )
-            return data[i] + X * correction[:, np.newaxis]
+            additional = rng.uniform(-self.bandwidth, self.bandwidth, size=size)
```

Comment:
I don't think this change is correct: it looks like this replaces the existing spherical tophat sampling (i.e. sampling from a hypersphere) with a cartesian tophat sampling (i.e. sampling from a hypercube). This is the same in 1 dimension, but differs in more than 1 dimension.

Reply:
Ah right, I had assumed the generalization to the multivariate case was a hypercube. It seems like if each dimension is distributed U(-b, b) then you end up with a hypercube, but a quick plot shows that isn't how the kernel function is implemented. Is there a benefit to having the kernel functions be rotationally symmetrical? The added complexity for the uniform kernel makes sense now. Is there a better way of implementing this for the other kernels (other than brute forcing with accept-reject or similar)? Maybe this could still be useful for the univariate case, or maybe I'll come back with a better implementation

Reply:
The benefit of a rotationally symmetric kernel is that it depends only on the distance between points rather than the angle/orientation. This is typical of kernels used in multiple dimensions, and has the added benefit that the KDE can be computed quickly using a fast tree-based implementation.

As for implementing the other samplers: I'd avoid accept/reject approaches  because of the curse of dimensionality. I'm certain there exist fast algorithms for the unimplemented kernels that are along the lines of the one used for the uniform case, but I don't know of any references.